### PR TITLE
copr: make copr task more concurrent

### PR DESCRIPTION
**Pull request by BusyJay**
Currently tasks in a batch of the same regions will be executed sequentially, we can handle it concurrently.

@siddontang @disksing PTAL

### CODE REVIEW

**Comment by ngaut** [comment]
Any benchmark results?

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 1]
```diff
@@ -24,25 +24,32 @@ use raftstore::Result;
...
 
     pub fn from_raw(db: Arc<DB>, region: Region) -> RegionSnapshot {
         RegionSnapshot {
-            snap: Snapshot::new(db),
+            snap: Snapshot::new(db).into_sync(),
             region: region,
         }
     }
 
+    pub fn clone(&self) -> RegionSnapshot {
```
Comment (siddontang) [inline src/raftstore/coprocessor/region_snapshot.rs]: can we derive `Clone` trait directly?

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 2]
```diff
@@ -24,25 +24,32 @@ use raftstore::Result;
 ///
 /// Only data within a region can be accessed.
 pub struct RegionSnapshot {
-    snap: Snapshot,
+    snap: SyncSnapshot,
     region: Region,
```
Comment (siddontang) [inline src/raftstore/coprocessor/region_snapshot.rs]: need using `Arc` for region?

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/raftstore/coprocessor/region_snapshot.rs (hunk 1 above)
Comment (BusyJay) [inline src/raftstore/coprocessor/region_snapshot.rs]: No, it can't compile.

File: src/raftstore/coprocessor/region_snapshot.rs (hunk 2 above)
Comment (BusyJay) [inline src/raftstore/coprocessor/region_snapshot.rs]: region is very small, it's ok to copy it every time.

Comment (siddontang) [inline src/raftstore/coprocessor/region_snapshot.rs]: Assume a region has 3 peers, the size is region ID + epoch + 3 * peer + start + end, so the least size may be 72 + start + end. Most of the time, the start and end key may be > 10 bytes, so a region may have 100 bytes, seem not very small.

**Comment by BusyJay** [comment]
According to sysbench, performance is decreased a little bit (2% ~ 3%), but it should be able to make requests waiting in a batch wait less time than the past.

**Comment by ngaut** [comment]
Could you file another issue to track the performance issue?

**Comment by siddontang** [comment]
@BusyJay 

I am confused that why the performance is decreased by 2%-3%, within a margin of error?

File: src/server/coprocessor/endpoint.rs [hunk 3]
```diff
@@ -215,20 +215,23 @@ impl BatchRunnable<Task> for Host {
                             continue;
                         }
                     };
-                    let end_point = TiDbEndPoint::new(snap);
+                    let len = reqs.len() as f64;
```
Comment (siddontang) [inline src/server/coprocessor/endpoint.rs]: seem that we can add two metrics here

1. How many requests in one batch? If most of the time, it is only 1, the change seems un-necessary. 
2. How many ranges one request scans? If the request needs to scan many ranges, do we need to scan concurrently?

Comment (ngaut) [inline src/server/coprocessor/endpoint.rs]: Good idea, this PR looks good to me generally.  A better choice is to separate the query by point query and complex query, we can improve the case later.

**Pull request merged**